crates/starknet-devnet-types/src/rpc/transactions.rs

```diff
@@ -415,20 +446,51 @@ impl From<&ResourceBoundsWrapper> for starknet_api::transaction::fields::ValidRe
 impl BroadcastedTransactionCommonV3 {
     /// Checks if total accumulated fee of resource_bounds for l1 is equal to 0 or for l2 is not
     /// zero
-    pub fn is_l1_gas_zero_or_l2_gas_not_zero(&self) -> bool {
-        let l2_is_not_zero = (self.resource_bounds.inner.l2_gas.max_amount as u128)
-            * self.resource_bounds.inner.l2_gas.max_price_per_unit
-            > 0;
-        let l1_is_zero = (self.resource_bounds.inner.l1_gas.max_amount as u128)
-            * self.resource_bounds.inner.l1_gas.max_price_per_unit
-            == 0;
-
-        l1_is_zero || l2_is_not_zero
+    pub fn are_gas_bounds_valid(&self) -> bool {
+        let ResourceBoundsMapping { l1_gas, l2_gas, l1_data_gas } = &self.resource_bounds.inner;
+        // valid set of resources are:
+        // 1. l1_gas > 0, l2_gas = 0, l1_data_gas = none
+        // 2. l1_gas > 0, l2_gas > 0, l1_data_gas > 0
+        match (l1_gas, l2_gas, l1_data_gas) {
+            // l1 > 0 and l2 > 0 and l1 data is none => invalid
+            (l1, l2, None)
+                if l1.max_amount > 0
+                    && l1.max_price_per_unit > 0
+                    && l2.max_amount > 0
+                    && l2.max_price_per_unit > 0 =>
```

Comment:
Shouldn't it be (my formatting not necessarily correct):
```suggestion
                if (l1.max_amount > 0 || l1.max_price_per_unit > 0)
                    && (l2.max_amount > 0 || l2.max_price_per_unit > 0) =>
```

Reply:
no because max_amount * max_price_per_unit can be 0 if max_amount or max_price_per_unit is 0